src/storage/src/types/sources.rs

```diff
@@ -1523,25 +1528,119 @@ impl Generator {
             Generator::Counter => DataEncodingInner::RowCodec(
                 RelationDesc::empty().with_column("counter", ScalarType::Int64.nullable(false)),
             ),
+            Generator::Auction => DataEncodingInner::RowCodec(
+                RelationDesc::empty()
+                    .with_column("table", ScalarType::String.nullable(false))
+                    .with_column(
+                        "row_data",
+                        ScalarType::List {
+                            element_type: Box::new(ScalarType::String),
+                            custom_id: None,
+                        }
+                        .nullable(false),
+                    ),
+            ),
         };
         SourceDataEncoding::Single(DataEncoding::new(inner))
     }
 
-    pub fn by_offset(&self, offset: MzOffset) -> Row {
+    /// Returns the list of tables and their (column names, types) for use with
+    /// `CREATE VIEWS`. Returns empty if `CREATE VIEWS` should error.
+
+    pub fn views(&self) -> Vec<(&str, Vec<(&str, &str)>)> {
         match self {
-            Self::Counter => Row::pack_slice(&[Datum::Int64(offset.offset as i64)]),
+            Generator::Auction => {
+                vec![
+                    (
+                        "auctions",
+                        vec![
+                            ("id", "INT8"),
+                            ("item", "STRING"),
+                            ("end_time", "TIMESTAMPTZ"),
+                        ],
+                    ),
+                    (
+                        "bids",
+                        vec![
+                            ("id", "INT8"),
+                            ("auction_id", "INT8"),
+                            ("amount", "INT4"),
+                            ("bid_time", "TIMESTAMPTZ"),
+                        ],
+                    ),
+                ]
+            }
+            Generator::Counter => Vec::new(),
+        }
+    }
+
+    pub fn by_seed(&self, now: NowFn, seed: Option<u64>) -> GeneratorIterator {
+        GeneratorIterator {
+            generator: self.clone(),
+            invocation: 0,
+            pending: VecDeque::new(),
+            rng: SmallRng::seed_from_u64(seed.unwrap_or_default()),
+            now,
         }
     }
 }
 
-impl FromStr for Generator {
-    type Err = String;
+pub struct GeneratorIterator {
+    generator: Generator,
+    // Number of times `pending` has been refilled.
+    invocation: i64,
+    pending: VecDeque<Row>,
```

Comment:
This is my new idea for how generators should work: each invocation can produce many rows, that way the generator can make related rows if needed, then the source accesses the generator as an iterator.

Reply:
Will all the rows be relocked at the same timestamp? I'm wondering if this can be used to show off transactional results?

Reply:
They currently do not, but we could change them to be reclocked to the same timestamp if some load generator needed that, yes.

Reply:
I wouldn’t block this PR on that change, I’m just trying to imagine the possibilities